rasa/core/policies/rule_policy.py

```diff
@@ -415,40 +423,62 @@ def _check_prediction(
     ) -> Optional[Text]:
 
         predicted_action_name = self._predict_next_action(tracker, domain, interpreter)
-        if not predicted_action_name or predicted_action_name == gold_action_name:
-            return None
-
         # RulePolicy will always predict active_loop first,
         # but inside loop unhappy path there might be another action
-        if predicted_action_name == tracker.active_loop_name:
+        if (
+            predicted_action_name != gold_action_name
+            and predicted_action_name == tracker.active_loop_name
+        ):
             rasa.core.test.emulate_loop_rejection(tracker)
             predicted_action_name = self._predict_next_action(
                 tracker, domain, interpreter
             )
-            if not predicted_action_name or predicted_action_name == gold_action_name:
-                return None
+
+        if collect_sources:
+            # we need to remember which action should be predicted by the rule
+            # in order to correctly output the names of the contradicting rules
+            self.lookup[RULES_SOURCES][self._prediction_source].append(
+                [tracker.sender_id, gold_action_name]
+            )
+            return
+
+        if not predicted_action_name or predicted_action_name == gold_action_name:
+            return
 
         tracker_type = "rule" if tracker.is_rule_tracker else "story"
-        return (
+        contradicting_rules = set(
+            rule_name
+            for rule_name, action_name in self.lookup[RULES_SOURCES][
+                self._prediction_source
+            ]
+            if action_name != gold_action_name
+        )
+
+        error_message = (
             f"- the prediction of the action '{gold_action_name}' in {tracker_type} "
-            f"'{tracker.sender_id}' is contradicting with another rule or story."
+            f"'{tracker.sender_id}' "
+            f"is contradicting with rule(s) '{', '.join(contradicting_rules)}'"
         )
+        # outputting predicted action 'action_default_fallback' is confusing
+        if predicted_action_name != self._fallback_action_name:
+            error_message += f" which predicted action '{predicted_action_name}'"
 
-    def _find_contradicting_rules(
+        return error_message + "."
+
+    def _run_prediction_on_trackers(
         self,
         trackers: List[TrackerWithCachedStates],
         domain: Domain,
         interpreter: NaturalLanguageInterpreter,
-    ) -> None:
-        logger.debug("Started checking rules and stories for contradictions.")
-        # during training we run `predict_action_probabilities` to check for
-        # contradicting rules.
-        # We silent prediction debug to avoid too many logs during these checks.
-        logger_level = logger.level
-        logger.setLevel(logging.WARNING)
+        collect_sources: bool,
+    ) -> List[Text]:
+        if collect_sources:
+            self.lookup[RULES_SOURCES] = defaultdict(list)
```

Comment:
why do we store this in `self.lookup`? Let's just use a new class attribute of this instead.

Reply:
Then we will automatically persist it. And I think it can be a great origin for logging information for debugging rules during prediction. I thought of adding it in different PR, but can do it in this PR as well

Reply:
> Then we will automatically persist it. 

We can also persist it if it's a separate attribute. Storing everything in this dictionary seems very hacky.

Reply:
why would we create additional persistence of smth that is already persisted. However, it might be better to update original `self.lookup[RULES]` with origin of the rule in addition to the action name

Reply:
I renamed it to private variable `self._rules_sources`